### PR TITLE
Add Ironic Inspector support + Networking/DHCP changes

### PR DESCRIPTION
**Pull request by hjensas**
Adds the Ironic Inspector CRDs, and a controller to deploy inspector with PXE services for DHCP/TFTP/HTTPBOOT to a separate statefulset.

Adds 'NetworkAttachments' to Conductor CRD and utilize annotations from openstack-k8s-operators/lib-common to build the annotiations for pod network connectivity. With this change we can add multiple networks to conductor and inspector pods, for example a separate bmc-network.

```
ironicConductor:
  networkAttachments:
    - baremetal-net
    - bmc-net
  provisionNetwork: baremetal-net
ironicInspector:
    networkAttachments:
    - baremetal-net
    inspectionNetwork: baremetal-net
```

Also change/and rename DHCPRange property to DHCPRanges, it now take a list of DHCPRange and the DHCPRange type has been extended to to allow additional properties required for routed networks. DHCPRange also has a "podIndex" property which is used by init containers to filter, this allow specifying separate ranges for each statefulset pod which ensures no duplicate addresses are leased.

    dhcpRanges:
    - name: netA
      start: 192.168.24.100
      end: 192.168.24.149
      netmask: 255.255.255.0
      gateway: 192.168.24.254
      mtu: 1440
      podIndex: 0
    - name: netB
      start: 192.168.24.150
      end: 192.168.24.199
      netmask: 255.255.255.0
      gateway: 192.168.24.254
      mtu: 1440
      podIndex: 1

```

Network definitions used in test env:
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: baremetal-net
spec:
  nodeSelector:
    node-role.kubernetes.io/worker: ""
  config: |-
    {
      "cniVersion": "0.3.1",
      "name": "baremetal-net",
      "type": "macvlan",
      "master": "ens7",
      "ipam": {
        "type": "whereabouts",
        "range": "192.168.24.0/24",
        "exclude": [
          "192.168.24.16/28",
          "192.168.24.32/27",
          "192.168.24.64/26",
          "192.168.24.128/25"
        ]
      }
    }
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: bmc-net
spec:
  nodeSelector:
    node-role.kubernetes.io/worker: ""
  config: |-
    {
      "cniVersion": "0.3.1",
      "name": "bmc-net",
      "type": "macvlan",
      "master": "ens8",
      "ipam": {
        "type": "whereabouts",
        "range": "192.168.253.0/24",
        "exclude": [
          "192.168.253.1/32",
          "192.168.253.10/31",
          "192.168.253.12/30",
          "192.168.253.16/28",
          "192.168.253.32/27",
          "192.168.253.64/26",
          "192.168.253.128/25"
        ]
      }
    }

```

Sucessful node enroll and inspect:
```
$ openstack baremetal node list
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| UUID                                 | Name  | Instance UUID | Power State | Provisioning State | Maintenance |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| 56a7002a-8acf-474a-8606-6d09def896ff | node0 | None          | None        | enroll             | False       |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
$ openstack baremetal node manage node0
$ openstack baremetal node list
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| UUID                                 | Name  | Instance UUID | Power State | Provisioning State | Maintenance |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| 56a7002a-8acf-474a-8606-6d09def896ff | node0 | None          | None        | verifying          | False       |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
$ openstack baremetal node list
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| UUID                                 | Name  | Instance UUID | Power State | Provisioning State | Maintenance |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
| 56a7002a-8acf-474a-8606-6d09def896ff | node0 | None          | power off   | manageable         | False       |
+--------------------------------------+-------+---------------+-------------+--------------------+-------------+
$ openstack baremetal introspection start node0
$ openstack baremetal introspection list
+--------------------------------------+---------------------+-------------+-------+
| UUID                                 | Started at          | Finished at | Error |
+--------------------------------------+---------------------+-------------+-------+
| 56a7002a-8acf-474a-8606-6d09def896ff | 2023-01-31T12:44:42 | None        | None  |
+--------------------------------------+---------------------+-------------+-------+
$ openstack baremetal introspection list
+--------------------------------------+---------------------+---------------------+-------+
| UUID                                 | Started at          | Finished at         | Error |
+--------------------------------------+---------------------+---------------------+-------+
| 56a7002a-8acf-474a-8606-6d09def896ff | 2023-01-31T12:44:42 | 2023-01-31T12:47:56 | None  |
+--------------------------------------+---------------------+---------------------+-------+
```

### CODE REVIEW

**Comment by hjensas** [comment]
Following is the sample config I used to deploy.

```
apiVersion: ironic.openstack.org/v1beta1
kind: Ironic
metadata:
  name: ironic
  namespace: openstack
spec:
  serviceUser: ironic
  customServiceConfig: |
    [DEFAULT]
    debug = true
  databaseInstance: openstack
  databaseUser: ironic
  # rpcTransport: oslo
  ironicAPI:
    replicas: 1
    containerImage: quay.io/tripleomastercentos9/openstack-ironic-api:current-tripleo
  ironicConductor:
    replicas: 1
    containerImage: quay.io/tripleomastercentos9/openstack-ironic-conductor:current-tripleo
    pxeContainerImage: quay.io/tripleomastercentos9/openstack-ironic-pxe:current-tripleo
    tftpProxyContainerImage: quay.io/steveb/tftpget2http:main-latest
    storageRequest: 10G
    storageClass: standard-csi
    # storageClass: crc-csi-hostpath-provisioner
    networkAttachments:
    - baremetal-net
    - bmc-net
    provisionNetwork: baremetal-net
    dhcpRanges:
    - name: netA
      start: 192.168.24.100
      end: 192.168.24.149
      netmask: 255.255.255.0
      gateway: 192.168.24.254
      mtu: 1440
      # Index 0 is the default
      # podIndex: 0
    - name: netB
      start: 192.168.24.150
      end: 192.168.24.199
      netmask: 255.255.255.0
      gateway: 192.168.24.254
      mtu: 1440
      podIndex: 1
    # dhcpRanges
    # - start: '172.20.0.33'
    #   end: '172.20.0.50'
  ironicInspector:
    # debug:
    #   service: true
    replicas: 1
    containerImage: quay.io/tripleomastercentos9/openstack-ironic-inspector:current-tripleo
    pxeContainerImage: quay.io/tripleomastercentos9/openstack-ironic-pxe:current-tripleo
    databaseInstance: openstack
    networkAttachments:
    - baremetal-net
    inspectionNetwork: baremetal-net
    dhcpRanges:
    - name: netA
      start: 192.168.24.200
      end: 192.168.24.219
      netmask: 255.255.255.0
      gateway: 192.168.254.1
      mtu: 1440
      # Index 0 is the default
      # podIndex: 0 
    - name: netB
      start: 192.168.24.220
      end: 192.168.24.240
      netmask: 255.255.255.0
      gateway: 192.168.254.1
      mtu: 1440
      podIndex: 1
    #dhcpRanges:
    #- start: '192.168.24.200'
    #  end: '192.168.24.219'
    #  podIndex: 0
    #- start: '192.168.24.220'
    #  end: '192.168.24.239'
    #  podIndex: 1
    #dhcpRanges:
    #- start: '172.20.0.51'
    #  end: '172.20.0.59'
  secret: ironic-secret
```

File: pkg/ironicinspector/service.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return ""
+	}
+
+	if len(routeList.Items) == 0 {
+		return ""
+	}
+	hostname := routeList.Items[0].Spec.Host
+	return strings.Split(hostname, ".apps.")[1]
+}
```
Comment (steveb) [inline pkg/ironicinspector/service.go:102]: Maybe we could refactor this so its shared with ironicconductor/service.go. Or not, its mostly descriptive

File: templates/ironicinspector/bin/pxe-init.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+if [ ! -d "/var/lib/ironic/httpboot" ]; then
+    mkdir /var/lib/ironic/httpboot
+fi
+
+# DHCP server configuration
+export InspectorNetworkIP=$(/usr/local/bin/container-scripts/get_net_ip ${InspectionNetwork})
+export INSPECTOR_HTTP_URL=$(python3 -c 'import os; print(os.environ["InspectorHTTPURL"] % os.environ)')
+
+export DNSMASQ_CFG=/var/lib/config-data/merged/dnsmasq.conf
+export DNSMASQ_CFG=/var/lib/config-data/merged/dnsmasq.conf
```
Comment (steveb) [inline templates/ironicinspector/bin/pxe-init.sh:32]: duplicate line

File: controllers/ironic_controller.go [hunk 3]
```diff
@@ -729,3 +755,37 @@ func (r *IronicReconciler) transportURLCreateOrUpdate(instance *ironicv1.Ironic)
 
 	return transportURL, op, err
 }
+
+func (r *IronicReconciler) inspectorDeploymentCreateOrUpdate(
+	instance *ironicv1.Ironic,
+) (*ironicv1.IronicInspector, controllerutil.OperationResult, error) {
+	deployment := &ironicv1.IronicInspector{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-inspector", instance.Name),
```
Comment (steveb) [inline controllers/ironic_controller.go:764]: "inspector" could come from the component constant. Not a big deal

**Review by steveb — APPROVED** [review]
Lets land this as is, then iterate

/lgtm

**Pull request merged**